### PR TITLE
[overridable API] Added Tox API in a structure form.

### PR DESCRIPTION
**Pull request by yurivict**
This is useful to facilitate API subclassing.
Some library may want to overload several functions, keeping the rest, like this:

```
MyLibToxApi mylibtox_api = tox_api;
mylib_tox_api.tox_func1 = &mylibtox_func1;
mylib_tox_api.tox_func2 = &mylibtox_func2;
```

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/502)

### CODE REVIEW

**Comment by iphydf** [comment]
Can you add a comment about how this is supposed to be used? I don't expect users will search git history for this PR and the usage docs you wrote in the PR body.

File: toxcore/tox.h [hunk 1]
```diff
@@ -2933,6 +2933,167 @@ uint16_t tox_self_get_udp_port(const Tox *tox, TOX_ERR_GET_PORT *error);
...
+    bool (*tox_friend_send_lossless_packet)(Tox *tox, uint32_t friend_number, const uint8_t *data, size_t length,
+                                            TOX_ERR_FRIEND_CUSTOM_PACKET *error);
+    void (*tox_callback_friend_lossy_packet)(Tox *tox, tox_friend_lossy_packet_cb *callback);
+    void (*tox_callback_friend_lossless_packet)(Tox *tox, tox_friend_lossless_packet_cb *callback);
+    void (*tox_self_get_dht_id)(const Tox *tox, uint8_t *dht_id);
+    uint16_t (*tox_self_get_udp_port)(const Tox *tox, TOX_ERR_GET_PORT *error);
+    uint16_t (*tox_self_get_tcp_port)(const Tox *tox, TOX_ERR_GET_PORT *error);
+} ToxApi;
+
+extern ToxApi tox_api;
```
Comment (iphydf) [inline toxcore/tox.h]: Are users supposed to change this global mutable variable?

Comment (yurivict) [inline toxcore/tox.h]: No, users are supposed to copy it to create a flavor of the API based on the original API.

**Comment by yurivict** [comment]
Made variables const.

**Comment by nurupo** [comment]
Uh, do we really want toxcore to provide this? I have yet to see a library that does this.

Also, the structhe names are shortened for some reason. `tox_api` vs `toxcore_api`, same with toxencryptsave.

**Comment by yurivict** [comment]
@nurupo I am going to create suchmid-level library.
Matue API should provide this for robustness.
I will fix structure names.

**Comment by yurivict** [comment]
I fixed the structure names.

**Comment by sudden6** [comment]
Why is this better than just calling the original Tox functions from a library on top and forward all the parameters + return value?

**Comment by yurivict** [comment]
@sudden6 I don't understand your suggestion. Could you elaborate?

tox_friend_send_message will be in both toxcore and library. How would you call it?

**Comment by sudden6** [comment]
I'd overwrite it with my own libraries prefix like `supertox_friend_send_message`

**Comment by yurivict** [comment]
> I'd overwrite it with my own libraries prefix like supertox_friend_send_message

Then you will have to provide all ~150 functions, when with this approach you only need to provide what you actually need.

**Comment by sudden6** [comment]
I understand the purpose now, thanks to https://github.com/qTox/qTox/pull/4238#issuecomment-284564113

This doesn't seem to build right now.

**Comment by yurivict** [comment]
It doesn't build? What's the error?

**Comment by Diadlo** [comment]
@yurivict https://travis-ci.org/TokTok/c-toxcore/jobs/208336424#L1709

**Comment by yurivict** [comment]
Please merge this for the tox defragmenter: https://github.com/TokTok/c-toxcore/issues/492#issuecomment-285888816

**Comment by iphydf** [comment]
I'm reluctant to merge this. I see no reason why you couldn't have this struct in a separate library. What other libraries provide this kind of overriding facility?

**Comment by yurivict** [comment]
Please merge this PR.

qTox has plugins system planned. I would like to add the type of plugin that will allow to alter the API. One application is previously proposed by me arbitrary size messaging, which I think will fit well as a plugin. But API needs to be alterable.

**Comment by Diadlo** [comment]
@yurivict In your case, you need to reimplement qTox `Core`. I don't think, that there is really reason to change core API. Just use wrapper

**Comment by yurivict** [comment]
> In your case, you need to reimplement qTox Core. I don't think, that there is really reason to change toxcore API.

The reasons:
1. If other clients were to develop advanced features like plugins, etc., they would need this in the same way as qTox.
2. This belongs near to where the API is defined, in toxcore, and not in clients.

**Comment by Diadlo** [comment]
@yurivict Change toxcore API it's wrong way to do it. You can create your wrapper and force all(?) clients to use it. But client should always give an avility to use toxcore directly
I'm against of this changes in toxcore

**Comment by Diadlo** [comment]
@yurivict You can write you library, which will provide few function line `yurivict_tox_friend_send_message` and custom callback setters. Anyone can integrate it by using your functions => profit

**Comment by yurivict** [comment]
@Diadlo Plugin should be generic, and shouldn't operate in terms of individual functions. It can operate through the interface:

```
namespace QToxPlugin
{
class ApiInterface : public QObject
{
    Q_OBJECT
public:
    const ToxavApi* overrideApi(const ToxavApi* prevApi);
};
};
```

This way it is generic, expandable, and doesn't have to implement a few functions here and there.

**Comment by Diadlo** [comment]
You may don't believe, but you still need "to implement a few functions here and there" to use it inside child classes to override toxcore function

**Note:** What if someone don't whant to use Qt or C++? This way not enough generic from this side.

**Comment by yurivict** [comment]
Other languages have other ways to deal with such things. APIs are normally defined in C and later other languages bind to them in their own way. And no, C apps can use it as is, without defining anything else.

**Comment by iphydf** [comment]
The checkbox allowing maintainers to rebase is not checked. We can't operate on this PR as is, and I'm not happy with the PR itself. I'm closing it for now. We may revisit it in the future.

**Pull request closed without merging**